File: android/jvm-tests/src/test/kotlin/SyncCryptoTest.kt
import hu.breaker.app.core.SyncCrypto
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

/**
 * A szinkron titkosítása — és ami ennél is fontosabb: hogy UGYANAZ, mint a gépen.
 *
 * Az alábbi értékeket NEM ez a kód gyártotta: a Node oldal (a valódi asztali
 * kód) állította elő őket, és ide vannak másolva. Ha a Kotlin mag bármiben
 * elcsúszik — kulcsszármaztatás, HKDF-címke, blob-formátum, base64 —, akkor ezek
 * a burkolatok nem nyílnak ki, és a teszt elbukik. Enélkül a hiba csak ott
 * derülne ki, ahol a legrosszabb: a felhasználó telefonján, belépéskor.
 */
class SyncCryptoTest {

    private val accountId = "acc_teszt"
    private val password = "ez-egy-elég-hosszú-jelszó"
    private val recoveryCode = "1THW-EBEP-44H9-33JC-EP8S-0S7Q-DVTF-CD6Q"
    private val wrappedByPassword = "brk1.ifgZSdAaT1J8Ijg3.a_WINPyrRBkB_ESkOI3cFg.Z9ftmx5XrbXx_bzxB9ceUFhysScGHpCJUU5AnMG1eHBVtQQJCZ8JOgFxBCc"
    private val wrappedByRecovery = "brk1.LYoCmyhyaS6jd-CB.vNAQ2qd5Z5F5_649M-q-7A.xeEMppLrHg-J2ytfrDieBEV3KOvLpJILtAmwnmifrH5F1tvUUdOu25-cQWY"
    private val dataKeyB64 = "nsAQTtKsDABPQkyrFumULr0hngffP/plHMWKEbBkifQ="
    private val blobFromNode = "brk1.FojyZdtJYV03Fzh1.V8csd6RWCZUHcg2aT-ij4A.63ep7GMAD9KX7kwNOoMERrLzHaK4imI4mU-2fII2lweHaPa3RMIXIA"

    @Test
    fun `the password opens a data key wrapped on the desktop`() {
        val key = SyncCrypto.unlockWithPassword(accountId, password, wrappedByPassword)
        assertEquals(dataKeyB64, SyncCrypto.b64(key))
    }

    @Test
    fun `the recovery code opens the same data key`() {
        val key = SyncCrypto.unlockWithRecovery(recoveryCode, wrappedByRecovery)
        assertEquals(dataKeyB64, SyncCrypto.b64(key))
        // Kötőjelek és kisbetűk nélkül is: kézzel másolva senki nem figyel ezekre.
        val messy = recoveryCode.lowercase().replace("-", " ")
        assertEquals(dataKeyB64, SyncCrypto.b64(SyncCrypto.unlockWithRecovery(messy, wrappedByRecovery)))
    }

    @Test
    fun `a blob written on the desktop decrypts here`() {
        val key = SyncCrypto.unb64(dataKeyB64)
        assertEquals("""[{"id":"site_1","domain":"youtube.com"}]""", SyncCrypto.decrypt(key, blobFromNode))
    }

    @Test
    fun `the auth keys match the ones the desktop sends`() {
        assertEquals("SPgYkvURBdAHhJqHf6fDAAWoPBMxS3oqvosFvdPTU0w=", SyncCrypto.authKey(password, accountId))
        assertEquals("tzOn+uxOlbCzUTlK0caIOxaRH3/pNxf+Yeq7826wlFI=", SyncCrypto.recoveryAuthKey(recoveryCode))
    }

    @Test
    fun `what we encrypt here can be read back here`() {
        val key = SyncCrypto.unb64(dataKeyB64)
        val a = SyncCrypto.encrypt(key, "youtube.com")
        val b = SyncCrypto.encrypt(key, "youtube.com")
        assertTrue(a != b, "azonos IV ugyanazzal a kulccsal GCM-nél a kulcs eldobása lenne")
        assertEquals("youtube.com", SyncCrypto.decrypt(key, a))
        assertEquals("youtube.com", SyncCrypto.decrypt(key, b))
    }

    @Test
    fun `a tampered blob is refused, not silently accepted`() {
        val key = SyncCrypto.unb64(dataKeyB64)
        val parts = blobFromNode.split(".")
        val ct = SyncCrypto.unb64(parts[3])
        ct[0] = (ct[0].toInt() xor 1).toByte()
        val tampered = listOf(parts[0], parts[1], parts[2], SyncCrypto.b64(ct).replace("+", "-").replace("/", "_").replace("=", ""))
            .joinToString(".")
        assertFailsWith<Exception> { SyncCrypto.decrypt(key, tampered) }
        assertFailsWith<IllegalArgumentException> { SyncCrypto.decrypt(key, "csak.harom.resz") }
        assertFailsWith<IllegalArgumentException> { SyncCrypto.decrypt(key, blobFromNode.replace("brk1", "brk9")) }
    }

    @Test
    fun `a wrong password does not open the wrapper`() {
        assertFailsWith<Exception> {
            SyncCrypto.unlockWithPassword(accountId, password + "x", wrappedByPassword)
        }
    }

    @Test
    fun `look-alike letters in the recovery code land in the same place`() {
        assertEquals("0111", SyncCrypto.normalizeRecoveryCode("o1-Il"))
        assertFailsWith<IllegalArgumentException> { SyncCrypto.recoveryKey("rövid") }
    }
}

File: desktop/test/sync-crypto.test.ts
// A szinkron titkosítása.
//
// Amit ezek a tesztek őriznek, az egyetlen mondat: a KISZOLGÁLÓ NEM LÁTJA az
// adatot. Ha ez elromlik, a felhasználó blokklistája és a mért ideje — vagyis
// pont az, amivel küzd — egy idegen gépen fekszik olvashatóan.
//
// A scrypt szándékosan lassú (32 MB, pár tized másodperc), ezért itt kevés
// kulcsszármaztatás van, és amit lehet, újrahasznosítunk.

import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import {
  BLOB_PREFIX, decrypt, encrypt, enroll, newRecoveryCode, normalizeRecoveryCode,
  recoveryKey, rewrapForNewPassword, unlockWithPassword, unlockWithRecovery,
  unwrapDataKey, wrapDataKey,
} from '../src/shared/sync/crypto';

const ACCOUNT = 'acc_teszt';
const PASSWORD = 'ez-egy-elég-hosszú-jelszó';

// Egyszer állítjuk elő, mert a scrypt szándékosan drága.
const enrollment = enroll(ACCOUNT, PASSWORD);

test('what the server gets never contains the data key', () => {
  const onServer = JSON.stringify(enrollment.serverSide);
  assert.ok(!onServer.includes(enrollment.dataKey.toString('base64')),
    'az adatkulcs nyersen sosem kerülhet a kiszolgálóra');
  assert.ok(!onServer.includes(PASSWORD), 'a jelszó sem');
  assert.ok(!onServer.includes(enrollment.recoveryCode), 'a helyreállító kód sem');
  assert.equal(enrollment.serverSide.accountId, ACCOUNT);
});

test('the password opens the data key, a wrong one does not', () => {
  const key = unlockWithPassword(ACCOUNT, PASSWORD, enrollment.serverSide.wrappedByPassword);
  assert.deepEqual(key, enrollment.dataKey);
  assert.throws(
    () => unlockWithPassword(ACCOUNT, PASSWORD + 'x', enrollment.serverSide.wrappedByPassword),
    'rossz jelszóval nem szabad kicsomagolni',
  );
});

test('the recovery code opens the same data key', () => {
  const key = unlockWithRecovery(enrollment.recoveryCode, enrollment.serverSide.wrappedByRecovery);
  assert.deepEqual(key, enrollment.dataKey, 'ugyanaz az adatkulcs, másik ajtón');
  // Kötőjelek és kisbetűk nélkül is: kézzel másolva senki nem figyel ezekre.
  const messy = enrollment.recoveryCode.toLowerCase().replace(/-/g, ' ');
  assert.deepEqual(
    unlockWithRecovery(messy, enrollment.serverSide.wrappedByRecovery),
    enrollment.dataKey,
  );
});

test('changing the password keeps the data readable', () => {
  // Ez a lényege annak, hogy az adatkulcs véletlen, nem a jelszóból származik:
  // a kiszolgáló nem tudná újratitkosítani az adatot, mert nem látja.
  const blob = encrypt(enrollment.dataKey, 'youtube.com');
  const next = rewrapForNewPassword(ACCOUNT, enrollment.dataKey, 'másik-jelszó-lett');
  const key = unlockWithPassword(ACCOUNT, 'másik-jelszó-lett', next.wrappedByPassword);
  assert.equal(decrypt(key, blob), 'youtube.com');
  assert.notEqual(next.authKey, enrollment.serverSide.authKey, 'a belépőkulcs viszont változik');
});

test('a tampered blob is refused, not silently accepted', () => {
  // GCM-címke nélkül a kiszolgáló (vagy aki a helyére áll) bitenként babrálhatna
  // a blokklistán úgy, hogy észre se vesszük.
  const blob = encrypt(enrollment.dataKey, '{"sites":["youtube.com"]}');
  const [tag, iv, auth, ct] = blob.split('.');
  assert.equal(tag, BLOB_PREFIX);

  const flipped = Buffer.from(ct, 'base64url');
  flipped[0] ^= 0x01;
  assert.throws(() => decrypt(enrollment.dataKey, [tag, iv, auth, flipped.toString('base64url')].join('.')));

  assert.throws(() => decrypt(enrollment.dataKey, blob.replace(BLOB_PREFIX, 'brk9')),
    'ismeretlen formátum');
  assert.throws(() => decrypt(enrollment.dataKey, 'csak.harom.resz'));
  assert.throws(() => decrypt(enrollment.dataKey, [tag, 'AAAA', auth, ct].join('.')),
    'rossz méretű IV');
});

test('the same text encrypts differently every time', () => {
  // Azonos IV ugyanazzal a kulccsal GCM-nél nem „gyengébb titkosítás”, hanem a
  // kulcs eldobása. Ha ez a teszt elbukik, a véletlen IV veszett el.
  const a = encrypt(enrollment.dataKey, 'youtube.com');
  const b = encrypt(enrollment.dataKey, 'youtube.com');
  assert.notEqual(a, b);
  assert.equal(decrypt(enrollment.dataKey, a), decrypt(enrollment.dataKey, b));
});

test('a recovery code is long, readable, and free of look-alike letters', () => {
  const code = newRecoveryCode();
  assert.match(code, /^[0-9A-HJKMNP-TV-Z]{4}(-[0-9A-HJKMNP-TV-Z]{4}){7}$/,
    'nyolc négyes csoport, I/L/O/U nélkül');
  assert.notEqual(code, newRecoveryCode(), 'minden kód friss');
  assert.equal(normalizeRecoveryCode('o1-Il'), '0111', 'a hasonló alakú jelek egy helyre esnek');
  assert.throws(() => recoveryKey('rövid'), 'a csonka kódot nem fogadjuk el');
});

test('a wrapped key of the wrong size is refused', () => {
  const kek = recoveryKey(newRecoveryCode());
  const bogus = wrapDataKey(kek, Buffer.alloc(8, 7)); // nem 32 bájt
  assert.throws(() => unwrapDataKey(kek, bogus), 'a méret ellenőrizve');
});

test('a short password cannot open an account', () => {
  // A jelszó itt nem egy weboldal belépője: EZ tartja a kulcsot, ami az adatot
  // nyitja. Aki a kiszolgálóra betör, offline próbálkozhat vele, korlátlanul —
  // ott már csak az scrypt lassúsága és a jelszó hossza védi.
  assert.throws(() => enroll('acc_rovid', 'rovid'), /legalább 10/);
  assert.throws(() => enroll('acc_rovid', ''), /legalább 10/);
  assert.throws(
    () => rewrapForNewPassword(ACCOUNT, enrollment.dataKey, 'rovid'),
    /legalább 10/,
    'jelszócserénél sem lehet gyengíteni',
  );
});
